dttools/src/list.c

```diff
@@ -1,406 +1,633 @@
 /*
-Copyright (C) 2003-2004 Douglas Thain and the University of Wisconsin
-Copyright (C) 2005- The University of Notre Dame
+Copyright (C) 2018- The University of Notre Dame
 This software is distributed under the GNU General Public License.
 See the file COPYING for details.
 */
 
+#include <assert.h>
+#include <limits.h>
+#include <math.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
 #include "list.h"
 
-#include <stdlib.h>
+struct list {
+	unsigned refcount;
+	unsigned length;
+	struct list_item *head;
+	struct list_item *tail;
+	struct list_cursor *iter;
+};
+
+struct list_item {
+	unsigned refcount;
```

Comment:
I'm not sure I like refcounts also for items. It will get messy when two cursors try to modify the list structure at the same time. 
If modifying the list is important, we can have a unique writing cursor. A list may have only one such cursor.

Reply:
So switching over to Rust's ownership model? :) I wouldn't mind having a `list_cursor_mut`, and it would simplify book keeping